EclipseLink IndexOutOfBoundsException: Index 0 out of bounds for length 0

After #2414 is fixed by beta 09, one of the tests that was previously blocked by that error, now fails with

```
Exception [EclipseLink-6168] (Eclipse Persistence Services - 5.0.0-B09.v202507111153-40df9e5d15636cd7801cd5f26058b0963b6496e5): org.eclipse.persistence.exceptions.QueryException
Exception Description: Query failed to prepare, unexpected error occurred: [java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0].
Internal Exception: java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
Query: ReadAllQuery(referenceClass=TaxPayer jpql="SELECT o FROM TaxPayer o WHERE (o.bankAccounts IS NOT EMPTY) ORDER BY o.ssn")
at org.eclipse.persistence.exceptions.QueryException.prepareFailed(QueryException.java:1617)
at org.eclipse.persistence.queries.DatabaseQuery.checkPrepare(DatabaseQuery.java:715)
at org.eclipse.persistence.queries.ObjectLevelReadQuery.checkPrepare(ObjectLevelReadQuery.java:1031)
at org.eclipse.persistence.queries.DatabaseQuery.checkPrepare(DatabaseQuery.java:645)
at org.eclipse.persistence.internal.jpa.EJBQueryImpl.buildEJBQLDatabaseQuery(EJBQueryImpl.java:246)
at org.eclipse.persistence.internal.jpa.EJBQueryImpl.<init>(EJBQueryImpl.java:132)
at org.eclipse.persistence.internal.jpa.EJBQueryImpl.<init>(EJBQueryImpl.java:112)
at org.eclipse.persistence.internal.jpa.EJBQueryImpl.<init>(EJBQueryImpl.java:92)
at org.eclipse.persistence.internal.jpa.EntityManagerImpl.createQuery(EntityManagerImpl.java:1846)
at io.openliberty.data.internal.persistence.QueryInfo.find(QueryInfo.java:1931)
at io.openliberty.data.internal.persistence.QueryInfo.find(QueryInfo.java:1858)
at io.openliberty.data.internal.persistence.RepositoryImpl.invoke(RepositoryImpl.java:567)
at jdk.proxy13/jdk.proxy13.$Proxy87.findByBankAccountsNotEmpty(Unknown Source)
at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
at org.jboss.weld.bean.proxy.AbstractBeanInstance.invoke(AbstractBeanInstance.java:39)
at org.jboss.weld.bean.proxy.ProxyMethodHandler.invoke(ProxyMethodHandler.java:109)
at test.jakarta.data.jpa.web.TaxPayers$2142678005$Proxy$_$$_WeldClientProxy.findByBankAccountsNotEmpty(Unknown Source)
at test.jakarta.data.jpa.web.DataJPATestServlet.testEmbeddableCollection(DataJPATestServlet.java:1445)
at ...
Caused by: java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:100)
at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:106)
at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:302)
at java.base/java.util.Objects.checkIndex(Objects.java:385)
at java.base/java.util.ArrayList.get(ArrayList.java:427)
at org.eclipse.persistence.internal.expressions.FunctionExpression.prepareObjectAttributeCount(FunctionExpression.java:837)
at org.eclipse.persistence.internal.expressions.FunctionExpression.normalize(FunctionExpression.java:423)
at org.eclipse.persistence.internal.expressions.SQLSelectStatement.normalize(SQLSelectStatement.java:1557)
at org.eclipse.persistence.internal.expressions.SubSelectExpression.normalizeSubSelect(SubSelectExpression.java:215)
at org.eclipse.persistence.internal.expressions.ExpressionNormalizer.normalizeSubSelects(ExpressionNormalizer.java:119)
at org.eclipse.persistence.internal.expressions.SQLSelectStatement.normalize(SQLSelectStatement.java:1630)
at org.eclipse.persistence.internal.queries.ExpressionQueryMechanism.buildNormalSelectStatement(ExpressionQueryMechanism.java:611)
at org.eclipse.persistence.internal.queries.ExpressionQueryMechanism.prepareSelectAllRows(ExpressionQueryMechanism.java:1808)
at org.eclipse.persistence.queries.ReadAllQuery.prepareSelectAllRows(ReadAllQuery.java:910)
at org.eclipse.persistence.queries.ReadAllQuery.prepare(ReadAllQuery.java:841)
at org.eclipse.persistence.queries.DatabaseQuery.checkPrepare(DatabaseQuery.java:696) 
```

The JPQL query is:
```
SELECT o FROM TaxPayer o WHERE (o.bankAccounts IS NOT EMPTY) ORDER BY o.ssn
```

Here is the Entity:
```
@Entity
public class TaxPayer {
    public enum FilingStatus {
        Single, MarriedFilingJointly, MarriedFilingSeparately, HeadOfHousehold
    }

    @ElementCollection
    public Set<AccountId> bankAccounts;

    @Basic(optional = false)
    public FilingStatus filingStatus;

    public float income;

    public int numDependents;

    @Id
    public long ssn;

    public TaxPayer() {
    }

    TaxPayer(long ssn, FilingStatus status, int numDependents, float income, AccountId... bankAccounts) {
        this.ssn = ssn;
        this.filingStatus = status;
        this.numDependents = numDependents;
        this.income = income;
        this.bankAccounts = new LinkedHashSet<AccountId>();
        for (AccountId account : bankAccounts)
            this.bankAccounts.add(account);
    }
}
```

The embeddable is:
```
@Embeddable
public class AccountId {

    public long accountNum;
    public long routingNum;

    public AccountId() {
    }

    public AccountId(long accountNum, long routingNum) {
        this.accountNum = accountNum;
        this.routingNum = routingNum;
    }

    public static AccountId of(long accountNum, long routingNum) {
        return new AccountId(accountNum, routingNum);
    }

    @Override
    public String toString() {
        return "AccountId:" + accountNum + ":" + routingNum;
    }
}
```

--- a/foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/expressions/FunctionExpression.java
+++ b/foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/expressions/FunctionExpression.java
@@ -834,7 +834,13 @@ public class FunctionExpression extends BaseExpression {
                 if ((newDescriptor.getPrimaryKeyFields().size() == 1) || !distinctUsed) {
                     // case 1: single PK =>
                     // treat COUNT(entity) as COUNT(entity.pk)
-                    Expression countArg = baseExp.getField(newDescriptor.getPrimaryKeyFields().get(0));
+                    Expression countArg = null;
+                    if (newDescriptor.getPrimaryKeyFields().size() > 1 && !newDescriptor.getAdditionalAggregateCollectionKeyFields().isEmpty()) {
+                        // count foreign key if primary is not specified like @ElementCollection with @Embeddable
+                        countArg = baseExp.getField(newDescriptor.getAdditionalAggregateCollectionKeyFields().get(0));
+                    } else {
+                        countArg = baseExp.getField(newDescriptor.getPrimaryKeyFields().get(0));
+                    }
                     if (distinctUsed) {
                         countArg = countArg.distinct();
                     }

--- a/foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/expressions/SubSelectExpression.java
+++ b/foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/expressions/SubSelectExpression.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 1998, 2024 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1998, 2025 Oracle and/or its affiliates. All rights reserved.
  * Copyright (c) 2022 IBM Corporation. All rights reserved.
  *
  * This program and the accompanying materials are made available under the
@@ -99,10 +99,12 @@ public class SubSelectExpression extends BaseExpression {
                 ClassDescriptor descriptor = baseExpression.getSession().getDescriptor(sourceClass);
                 if (descriptor != null){
                     DatabaseMapping mapping = descriptor.getMappingForAttributeName(attribute);
-                    if (mapping != null && mapping.isDirectCollectionMapping()){
+                    if (mapping != null && (mapping.isDirectCollectionMapping() || mapping.isAggregateCollectionMapping())){
                         subQuery.setExpressionBuilder(baseExpression.getBuilder());
                         subQuery.setReferenceClass(sourceClass);
-                        subQuery.addCount(attribute, subQuery.getExpressionBuilder().anyOf(attribute), returnType);
+                        if (mapping.isDirectCollectionMapping() || subQuery.getItems().isEmpty()) {
+                            subQuery.addCount(attribute, subQuery.getExpressionBuilder().anyOf(attribute), returnType);
+                        }
                         return;
                     }
                 }

--- a/jpa/eclipselink.jpa.testapps/jpa.test.jpql/src/test/java/org/eclipse/persistence/testing/tests/jpa/jpql/relationships/AdvancedQueryTest.java
+++ b/jpa/eclipselink.jpa.testapps/jpa.test.jpql/src/test/java/org/eclipse/persistence/testing/tests/jpa/jpql/relationships/AdvancedQueryTest.java
@@ -84,6 +84,7 @@ public class AdvancedQueryTest extends JUnitTestCase {
         suite.addTest(new AdvancedQueryTest("testConcurrentLoadGroup"));
         suite.addTest(new AdvancedQueryTest("testSelectOnNames"));
         suite.addTest(new AdvancedQueryTest("testSelectOnRecipientInfo"));
+        suite.addTest(new AdvancedQueryTest("testSelectOnEmbeddable"));
         suite.addTest(new AdvancedQueryTest("testTearDown"));
         return suite;
     }
@@ -284,7 +285,7 @@ public class AdvancedQueryTest extends JUnitTestCase {
         }
     }
 
-    // Issue #2414 test: IS NOT EMPTY on @ElementCollection of @Embeddable class
+    // Issue #2414 test: IS NOT EMPTY on @ElementCollection of @Embeddable class with single value (String in this case)
     public void testSelectOnRecipientInfo() {
         EntityManager em = createEntityManager();
         try {
@@ -296,4 +297,15 @@ public class AdvancedQueryTest extends JUnitTestCase {
         }
     }
 
+    // Issue #2477 test: IS NOT EMPTY on @ElementCollection of @Embeddable class with multiple attributes
+    public void testSelectOnEmbeddable() {
+        EntityManager em = createEntityManager();
+        try {
+            List<?> result = em.createQuery("SELECT o FROM ShippingAddress o WHERE (o.alternateAddresses IS NOT EMPTY)")
+                    .getResultList();
+            assertEquals(1, result.size());
+        } finally {
+            closeEntityManager(em);
+        }
+    }
 }

--- a/jpa/eclipselink.jpa.testapps/jpa.test.relationships/src/main/java/org/eclipse/persistence/testing/models/jpa/relationships/AlternateAddress.java
+++ b/jpa/eclipselink.jpa.testapps/jpa.test.relationships/src/main/java/org/eclipse/persistence/testing/models/jpa/relationships/AlternateAddress.java
@@ -1,0 +1,52 @@
+/*
+ * Copyright (c) 2025 Oracle and/or its affiliates. All rights reserved.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0,
+ * or the Eclipse Distribution License v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+package org.eclipse.persistence.testing.models.jpa.relationships;
+
+import jakarta.persistence.ElementCollection;
+import jakarta.persistence.Embeddable;
+
+/**
+ * Alternate address.
+ * {@link Embeddable} class with {@link ElementCollection} attribute.
+ */
+@Embeddable
+public class AlternateAddress {
+
+    private int houseNumber;
+
+    private String streetName;
+
+    public AlternateAddress() {
+        this(-1, null);
+    }
+
+    public AlternateAddress(int houseNumber, String streetName) {
+        this.houseNumber = houseNumber;
+        this.streetName = streetName;
+    }
+
+    public int getHouseNumber() {
+        return houseNumber;
+    }
+
+    public void setHouseNumber(int houseNumber) {
+        this.houseNumber = houseNumber;
+    }
+
+    public String getStreetName() {
+        return streetName;
+    }
+
+    public void setStreetName(String streetName) {
+        this.streetName = streetName;
+    }
+}

--- a/jpa/eclipselink.jpa.testapps/jpa.test.relationships/src/main/java/org/eclipse/persistence/testing/models/jpa/relationships/RelationshipsExamples.java
+++ b/jpa/eclipselink.jpa.testapps/jpa.test.relationships/src/main/java/org/eclipse/persistence/testing/models/jpa/relationships/RelationshipsExamples.java
@@ -71,6 +71,7 @@ public class RelationshipsExamples {
         orderExample2.setSalesPerson(salesPerson2);
 
         ShippingAddress shippingAddressExample1 = shippingAddressExample1();
+        ShippingAddress shippingAddressExample2 = shippingAddressExample2();
 
         allObjects.add(customerExample1);
         allObjects.add(customerExample2);
@@ -95,6 +96,7 @@ public class RelationshipsExamples {
         allObjects.add(rep2);
 
         allObjects.add(shippingAddressExample1);
+        allObjects.add(shippingAddressExample2);
 
         UnitOfWork unitOfWork = session.acquireUnitOfWork();
         unitOfWork.registerAllObjects(allObjects);
@@ -205,7 +207,23 @@ public class RelationshipsExamples {
                                    List.of("Peter", "Parker"),
                                    new StreetAddress(2800,
                                                      "37th St NW",
-                                                     List.of("Receiving Dock", "Building 040-1")));
+                                                     List.of("Receiving Dock", "Building 040-1")),
+                                   null);
     }
 
+    public static ShippingAddress shippingAddressExample2() {
+        List<AlternateAddress> alternateAddresses = new ArrayList<>();
+        AlternateAddress alternateAddress1 = new AlternateAddress(456,"Kratka");
+        alternateAddresses.add(alternateAddress1);
+
+        return new ShippingAddress(2002L,
+                "Prague",
+                "Czech Republic",
+                11000,
+                null,
+                new StreetAddress(123,
+                        "Dlouha",
+                        null),
+                alternateAddresses);
+    }
 }

--- a/jpa/eclipselink.jpa.testapps/jpa.test.relationships/src/main/java/org/eclipse/persistence/testing/models/jpa/relationships/ShippingAddress.java
+++ b/jpa/eclipselink.jpa.testapps/jpa.test.relationships/src/main/java/org/eclipse/persistence/testing/models/jpa/relationships/ShippingAddress.java
@@ -42,8 +42,11 @@ public class ShippingAddress {
 
     private int zipCode;
 
+    @ElementCollection
+    List<AlternateAddress> alternateAddresses;
+
     public ShippingAddress() {
-        this(-1L, null, null, -1, List.of(), null);
+        this(-1L, null, null, -1, List.of(), null, List.of());
     }
 
     public ShippingAddress(Long id,
@@ -51,13 +54,16 @@ public class ShippingAddress {
                            String state,
                            int zipCode,
                            List<String> names,
-                           StreetAddress streetAddress) {
+                           StreetAddress streetAddress,
+                           List<AlternateAddress> alternateAddresses
+    ) {
         this.id = id;
         this.city = city;
         this.state = state;
         this.zipCode = zipCode;
         this.names = names;
         this.streetAddress = streetAddress;
+        this.alternateAddresses = alternateAddresses;
     }
 
     public Long getId() {
@@ -108,4 +114,11 @@ public class ShippingAddress {
         this.zipCode = zipCode;
     }
 
+    public List<AlternateAddress> getAlternateAddresses() {
+        return alternateAddresses;
+    }
+
+    public void setAlternateAddresses(List<AlternateAddress> alternateAddresses) {
+        this.alternateAddresses = alternateAddresses;
+    }
 }

